File: kotlin/src/main/kotlin/shared/ValueObjects.kt
package shared

import java.net.URI

/**
 * Value Objects: 不変の値オブジェクト群
 * Kotlinではvalue class (inline class)で型安全なラッパーを作成
 */

// ID型: value classでゼロコスト抽象化
@JvmInline
value class FrameworkId private constructor(val value: String) {
    companion object {
        fun create(value: String): Result<FrameworkId, ValidationError> =
            if (value.isBlank()) {
                err(ValidationError("id", "FrameworkId cannot be empty", "EMPTY_ID"))
            } else {
                ok(FrameworkId(value))
            }
    }
}

@JvmInline
value class ControlId private constructor(val value: String) {
    companion object {
        fun create(value: String): Result<ControlId, ValidationError> =
            if (value.isBlank()) {
                err(ValidationError("id", "ControlId cannot be empty", "EMPTY_ID"))
            } else {
                ok(ControlId(value))
            }
    }
}

@JvmInline
value class EvidenceId private constructor(val value: String) {
    companion object {
        fun create(value: String): Result<EvidenceId, ValidationError> =
            if (value.isBlank()) {
                err(ValidationError("id", "EvidenceId cannot be empty", "EMPTY_ID"))
            } else {
                ok(EvidenceId(value))
            }
    }
}

@JvmInline
value class RiskId private constructor(val value: String) {
    companion object {
        fun create(value: String): Result<RiskId, ValidationError> =
            if (value.isBlank()) {
                err(ValidationError("id", "RiskId cannot be empty", "EMPTY_ID"))
            } else {
                ok(RiskId(value))
            }
    }
}

@JvmInline
value class UserId private constructor(val value: String) {
    companion object {
        fun create(value: String): Result<UserId, ValidationError> =
            if (value.isBlank()) {
                err(ValidationError("id", "UserId cannot be empty", "EMPTY_ID"))
            } else {
                ok(UserId(value))
            }
    }
}

@JvmInline
value class IntegrationId private constructor(val value: String) {
    companion object {
        fun create(value: String): Result<IntegrationId, ValidationError> =
            if (value.isBlank()) {
                err(ValidationError("id", "IntegrationId cannot be empty", "EMPTY_ID"))
            } else {
                ok(IntegrationId(value))
            }
    }
}

/**
 * Percentage: 0-100の範囲を保証する値オブジェクト
 */
@JvmInline
value class Percentage private constructor(val value: Int) {
    companion object {
        fun create(value: Int): Result<Percentage, ValidationError> =
            if (value < 0 || value > 100) {
                err(ValidationError(
                    "percentage",
                    "Percentage must be between 0 and 100",
                    "INVALID_PERCENTAGE"
                ))
            } else {
                ok(Percentage(value))
            }
    }
}

/**
 * URL値オブジェクト
 */
@JvmInline
value class Url private constructor(val value: String) {
    companion object {
        fun create(value: String): Result<Url, ValidationError> =
            try {
                URI(value).toURL()
                ok(Url(value))
            } catch (e: Exception) {
                err(ValidationError("url", "Invalid URL format", "INVALID_URL"))
            }
    }
}

File: typescript/src/shared/value-objects.ts
/**
 * Value Objects: 不変の値オブジェクト群
 * TypeScriptでは Branded Types パターンで型安全なIDを実現
 */

import { Result, ok, err, ValidationError, validationError } from "./result.js";

// Branded Types: コンパイル時の型安全性を保証
declare const brand: unique symbol;
type Brand<T, B> = T & { readonly [brand]: B };

// ID型
export type FrameworkId = Brand<string, "FrameworkId">;
export type ControlId = Brand<string, "ControlId">;
export type EvidenceId = Brand<string, "EvidenceId">;
export type RiskId = Brand<string, "RiskId">;
export type UserId = Brand<string, "UserId">;
export type IntegrationId = Brand<string, "IntegrationId">;

// IDファクトリ関数
const createId = <T extends string>(
  value: string,
  typeName: string
): Result<Brand<string, T>, ValidationError> => {
  if (!value || value.trim().length === 0) {
    return err(validationError("id", `${typeName} cannot be empty`, "EMPTY_ID"));
  }
  return ok(value as Brand<string, T>);
};

export const frameworkId = (value: string): Result<FrameworkId, ValidationError> =>
  createId(value, "FrameworkId");

export const controlId = (value: string): Result<ControlId, ValidationError> =>
  createId(value, "ControlId");

export const evidenceId = (value: string): Result<EvidenceId, ValidationError> =>
  createId(value, "EvidenceId");

export const riskId = (value: string): Result<RiskId, ValidationError> =>
  createId(value, "RiskId");

export const userId = (value: string): Result<UserId, ValidationError> =>
  createId(value, "UserId");

export const integrationId = (value: string): Result<IntegrationId, ValidationError> =>
  createId(value, "IntegrationId");

/**
 * Percentage: 0-100の範囲を保証する値オブジェクト
 */
export type Percentage = Brand<number, "Percentage">;

export const percentage = (value: number): Result<Percentage, ValidationError> => {
  if (value < 0 || value > 100) {
    return err(
      validationError(
        "percentage",
        "Percentage must be between 0 and 100",
        "INVALID_PERCENTAGE"
      )
    );
  }
  return ok(value as Percentage);
};

/**
 * URL値オブジェクト
 */
export type Url = Brand<string, "Url">;

export const url = (value: string): Result<Url, ValidationError> => {
  try {
    new URL(value);
    return ok(value as Url);
  } catch {
    return err(validationError("url", "Invalid URL format", "INVALID_URL"));
  }
};
